File: Linked_List/141.Linked_list_cycle.py
class Solution:
    def hasCycle(self, head: ListNode) -> bool:
        if head is None or head.next is None:
            return False
        slow=head
        fast=head
        
        while fast.next!=None and fast.next.next!=None:
            slow=slow.next
            fast=fast.next.next
            if slow==fast:
                return True
        return False
        

File: Linked_List/Java/141. Linked List Cycle.java
#################################################Optimized solution############################################

public class Solution {
   public boolean hasCycle(ListNode head) {
      ListNode slow = head;
      ListNode fast = head;

      if (head == null || head.next == null) return false;

      while (fast.next != null && fast.next.next != null) {
         fast = fast.next.next;
         slow = slow.next;

         if (slow == fast) {
            return true;
         }

      }
      return false;
   }
}


#################################################Optimized solution############################################

public class Solution {
    public boolean hasCycle(ListNode head) {
       HashSet<ListNode> set=new HashSet<>();
       ListNode curr=head;
       while(curr!=null){
           if(set.contains(curr)) return true;
           set.add(curr);
           curr=curr.next;
       }
       return false;
    }
}